### DISCUSSION TITLE
Generic custom model

### DISCUSSION THREAD

**Discussion by benknab** [prisma/prisma#15868] [Q&A] [answered]
Is it possible to create a generic custom model? I want something similar that's described on the [custom models page](https://www.prisma.io/docs/concepts/components/prisma-client/custom-models), but I don't want to define a custom model for every Prisma model.

I would like to have this API, which would allow me to create a repository from a Prisma delegate and a mapper to map returned Prisma objects to my domain entities.

```ts
type User = {
	id: string
	email: string
}

type UserEntity = {
	// opaque/branded types in the entity
	id: Id
	email: Email
}

const toUserEntity = (user: User): UserEntity => {...}

const prisma = new PrismaClient()
const userRepository = new EntityRepository<UserEntity>(prisma.user, toUserEntity)

// `findMany` args is the same as the original Prisma `findMany` args, except `select` is omitted
const users = await userRepository.findMany({where: {email: {contains: "@gmail.com"}))
// `users` type is `UserEntity[]` instead of `User[]`
```

I can get this done just fine one by one, but I would like to have a generic repository. But I struggle to constrain the delegate to a single type when the repository is instantiated and the union breaks the arguments and return types:
```ts
class EntityRepository<Entity, Delegate extends PrismaClient["user"] | PrismaClient["transaction"]> {...}
```

How should I go about implementing this? Thanks for the help!

**Comment by benknab** [comment, marked as answer]
This does not seem possible with current Prisma types, see #3929 #5273 #7480

**Reply by nurul3101** [reply]
Hello @bkiac 👋 
Can you have a look at our Prisma Client Extensions Proposal: #15074 and let us know your thoughts, I am wondering if we could incorporate this through the proposal.

**Reply by benknab** [reply]
yeah, I think that proposal would help with my use case

I managed to create generic interfaces that I can extend/implement in my classes. Obviously, I would prefer if I could write the implementation in an abstract class or a parent class and inherit the generic Prisma methods but the interface works well enough for now.

(I'm removing `select` and `include` args because I don't want to allow the consumer to override what data to request otherwise I can't construct an entity)
```ts
type PrismaFindUniqueOrThrow = (args?: {where: any}) => any
export type FindUniqueOrThrowArgs<Fn extends PrismaFindUniqueOrThrow> = Pick<
  NonNullable<Parameters<Fn>[0]>,
  'where'
>
export type FindUniqueOrThrow<Fn extends PrismaFindUniqueOrThrow, Entity> = (
  args: FindUniqueOrThrowArgs<Fn>
) => Promise<Entity>

type PrismaFind = (args?: {select?: any; include?: any}) => any
export type FindArgs<Fn extends PrismaFind> = Omit<
  NonNullable<Parameters<Fn>[0]>,
  'select' | 'include'
>

export type FindFirstOrThrowArgs<Fn extends PrismaFind> = FindArgs<Fn>
export type FindFirstOrThrow<Fn extends PrismaFind, Entity> = (
  args: FindFirstOrThrowArgs<Fn>
) => Promise<Entity>

export type FindManyArgs<Fn extends PrismaFind> = FindArgs<Fn>
export type FindMany<Fn extends PrismaFind, Entity> = (
  args: FindManyArgs<Fn>
) => Promise<Entity[]>

export type PrismaClientMethod =
  | '$on'
  | '$connect'
  | '$disconnect'
  | '$use'
  | '$executeRaw'
  | '$executeRawUnsafe'
  | '$queryRaw'
  | '$queryRawUnsafe'
  | '$transaction'

export type PrismaModel = keyof Omit<PrismaClient, PrismaClientMethod>

export interface Repository<Model extends PrismaModel, Entity> {
  save: (entity: Entity) => Promise<Entity>
  findUniqueOrThrow: FindUniqueOrThrow<
    PrismaClient[Model]['findUniqueOrThrow'],
    Entity
  >
  findFirstOrThrow: FindFirstOrThrow<
    PrismaClient[Model]['findFirstOrThrow'],
    Entity
  >
  findMany: FindMany<PrismaClient[Model]['findMany'], Entity>
}
```

**Reply by nurul3101** [reply]
Thanks for sharing your solution, feel free to let us know your feedback once the Prisma Client Extensions is out 🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
